### ISSUE TITLE
Sometimes incoming & outgoing messages are duplicated

### ISSUE THREAD

**Issue by greatquux** [EionRobb/purple-teams#51]
Sometimes, not all the time, and I think usually when Pidgin has been open and working for some time, it will show all incoming and outgoing Teams messages twice.  They are only shown once in the Microsoft Teams client.  Closing and reopening fixes it.  (I should have tried disabling/enabling account but will have to do that next time.)  I captured a debug Pidgin log with the filter set to 'teams' for All messages when the problem was happening and will attach it in hopes it will be useful.

**Comment by greatquux** [comment]
[pidgin-dupe.log.txt](https://github.com/EionRobb/purple-teams/files/14605978/pidgin-dupe.log.txt)

**Comment by EionRobb** [comment]
Probably a surprising response from me, but yay! That's great that you're getting dupes!  That means the websocket bit is working well!

As a bit of backstory, I'm working on moving from the legacy http long-poll over to the websockets bit, but I feel like it's still a bit too unreliable to be the only method for messages, so was waiting for someone to open an issue to complain about the dupes haha.

I need to add in a de-duping system

**Comment by greatquux** [comment]
Well I certainly get that! But it's only happening sometimes, so I'm not sure if it really is reliable?

**Comment by EionRobb** [comment]
I improved the reliability of the websocket yesterday (it'd do silly things like stop receiving messages after a while, or not want to start receiving them if it wasn't the last event listener to be set up, how rude!) so this got more annoying.   I've now added in a buffer that checks if the incoming event is already within the last 10 events to be processed, and doesn't re-process if so - so hopefully the dupes have gone away now.

There's still an annoying thread-race-ish issue that when a new chat window is opened, it'll show the first message twice - because it's trying to fetch all the message history since the last known history point which excludes the new message, but then it also handles the new message coming through the event stream.

**Comment by greatquux** [comment]
I'll do some more testing though I am definitely still seeing it sometimes and seems to come and go:

![image](https://github.com/EionRobb/purple-teams/assets/1655942/5f0598e0-bbbc-4f9e-b47d-c886f7fc72a4)

**Comment by EionRobb** [comment]
Yeah, it's an improvement but it's still not ideal :/

Unfortunately there's not really a unique-enough field to use for incoming messages.  Currently using the message timestamp in nanoseconds but sometimes there's drift between the websocket and the http-poll streams that the exact same message comes through with two (microscopically) different timestamps

**Comment by greatquux** [comment]
But this is only temporary until you switch to websockets fully, right?  Perhaps an environment variable or some other config option can be created to pick only using one or the other, so we can test it out and see how reliable websockets are?

**Comment by EionRobb** [comment]
As of https://github.com/EionRobb/purple-teams/commit/a689c6cdda9bc9d3f775b1bcae6f38c7660d9a5f there's now a setting "only_use_websocket" (default off) that will skip the old poll way and only use the new websocket way, which should reduce dups

**Comment by greatquux** [comment]
only been testing for about 2 days with that option turned on but no missed messages or duplicates so far 😁

**Comment by EionRobb** [comment]
Nice one! Appreciate the feedback!

During that time have you noticed whether people are still having their online/away status updating?

**Comment by greatquux** [comment]
Yes it does look like that works fine - I checked a few people and also toggled my own status in the regular Teams client and it shows up pretty quickly in pidgin

**Comment by greatquux** [comment]
I am still seeing occasional duplicates even with the "only use websocket" option on, i think they are mostly confined to times when Pidgin is offline and attempting to get any messages it missed before.
